lib/remote_node.go

```diff
@@ -0,0 +1,648 @@
+}
+
+func (rn *RemoteNode) IsValidated() bool {
+	return rn.connectionStatus == RemoteNodeStatus_HandshakeCompleted
+}
+
+func (rn *RemoteNode) IsValidator() bool {
+	if !rn.IsValidated() {
+		return false
+	}
+	return rn.GetValidatorPublicKey() != nil
+}
+
+// DialOutboundConnection dials an outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialOutboundConnection(netAddr *wire.NetAddress) error {
+	if !rn.IsNotConnected() {
+		return fmt.Errorf("RemoteNode.DialOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// DialPersistentOutboundConnection dials a persistent outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialPersistentOutboundConnection(netAddr *wire.NetAddress) error {
+	if !rn.IsNotConnected() {
+		return fmt.Errorf("RemoteNode.DialPersistentOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialPersistentOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// ConnectInboundPeer connects a peer once a successful inbound connection has been established.
+func (rn *RemoteNode) ConnectInboundPeer(conn net.Conn, na *wire.NetAddress) error {
+	if !rn.IsNotConnected() {
+		return fmt.Errorf("RemoteNode.ConnectInboundPeer: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	id := rn.GetId().ToUint64()
+	rn.peer = rn.cmgr.ConnectPeer(id, conn, na, false, false)
+	rn.setStatusConnected()
+	return nil
+}
+
+// ConnectOutboundPeer connects a peer once a successful outbound connection has been established.
+func (rn *RemoteNode) ConnectOutboundPeer(conn net.Conn, na *wire.NetAddress, isPersistent bool) error {
+	if rn.connectionStatus != RemoteNodeStatus_Attempted {
+		return fmt.Errorf("RemoteNode.ConnectOutboundPeer: RemoteNode is not in the Attempted state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	id := rn.GetId().ToUint64()
+	rn.peer = rn.cmgr.ConnectPeer(id, conn, na, true, isPersistent)
+	rn.setStatusConnected()
+	return nil
+}
+
+// Disconnect disconnects the remote node, closing the attempted connection or the established connection.
+func (rn *RemoteNode) Disconnect() {
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	id := rn.GetId().ToUint64()
+	switch rn.connectionStatus {
+	case RemoteNodeStatus_Attempted:
+		rn.cmgr.CloseAttemptedConnection(id)
+	case RemoteNodeStatus_Connected, RemoteNodeStatus_HandshakeCompleted:
+		rn.cmgr.CloseConnection(id)
+	}
+	rn.setStatusTerminated()
+}
+
+func (rn *RemoteNode) SendMessage(desoMsg DeSoMessage) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected && rn.connectionStatus != RemoteNodeStatus_HandshakeCompleted {
+		return fmt.Errorf("SendMessage: Remote node is not connected")
+	}
```

Comment:
I'm a bit confused about the intended use of the `SendMessage` function.

Is the function intended to be called from outside of the RemoteNode struct? If so, you'll need the remote node to have completed the handshake. 

However, you call `SendMessage` for sending Version and Verack messages, in which case, you'll need to be able send the messages regardless of the handshake status of the remote node.

If my understanding above is correct, it seems best to use this `SendMessage` function for external uses only. And to have a private `sendMessage` function for internal uses
```suggestion
func (rn *RemoteNode) SendMessage(desoMsg DeSoMessage) error {
	if rn.connectionStatus != RemoteNodeStatus_HandshakeCompleted() {
		return fmt.Errorf("SendMessage: Remote node is not connected")
	}
	
	return rn.sendMessage(desoMsg)
}

func (rn *RemoteNode) sendMessage(desoMsg DeSoMessage) error {
```

Reply:
Nice simplification!